README.md
# Psylab
psylab is a collection of scripts, written in Matlab, for designing and controlling interactive psychoacoustical listening experiments in a uniform and quick manner.

Currently, $n$-AFC detection and discrimination experiments ($n=2,3,4$) and 2-AFC matching experiments are readily supported.  The stimulus variable can either be controlled adaptively according to $x$-up-$y$-down algorithms ($x,y \in \{1,2,3\})$) and weighted-up-down algorithms.  Alternatively, the method of constant stimuli may be used for detection and discrimination experiments.

psylab is based on many ideas of the programs 'SI/SISG' which were written at ``Drittes Physikalisches Institut'', Universität Göttingen, during the years 1989 to 1991, and which were distributed under GNU General Public License (GNU~GPL).

psylab is regularly used for education and research at Institut für Hörtechnik und Audiologie (IHA), Jade Hochschule in Oldenburg, Germany.

One aim of psylab has been to provide the students in the psychoacoustic courses with a uniform and simple starting point for designing one's own listening experiments.  So far, roughly +300 students at IHA have used psylab in their psychoacoustics assignments, for designing and carrying out individual listening experiments.  Their experience and comments have continuously contributed to new features and bug fixes of psylab.


ChangeLog
2016-11-04  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_check.m: added a check for correct value of M.MINSTEP,
	and for matching signs of M.STEP and M.MINSTEP

	corrected a typo in generation of m_postsig

2016-11-02  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_init_run.m: check/set default for new
	M.REVERSE_UP_AND_DOWN which is used for runs with reversed
	definition of "up" and "down" (see documentation).

	* mpsy_proto_conststim.m: new file for constant stimuli
	method, with functionality corresponding to mpsy_proto.m for
	adaptive methods.

	* mpsy_afc_conststim_main.m: implementation of the
	constant stimuli method for measuring percentage correct
	responses for a block-design.

	* mpsy_querysplash.m (mpsy_querysplash): new function that
	allows to bypass the user query in mpsy_init.m

	* mpsy_init.m: show a new (small) splash screen with info
	about license and no-warranty, and query user's OK.

	* psydat_helper.m (psydat_helper): reading of psydat file
	adapted to new version 3 psydat file format. 

	* mpsy_check_psydat_file.m (mpsy_check_psydat_file): new file
	name for previous file "mpsy_check_psydat_version.m".  Adapted
	functionality for new version3 psydat file format.

	* mpsy_proto_v2.m: new name *_v2 for old
	version 2 psydat file format

	* mpsy_proto.m: update to writing of the new
	version 3 psydat file format

	* read_psydat_v2.m (read_psydat_v2): new name *_v2 for old
	version 2 psydat file format

	* read_psydat.m (read_psydat): update to reading of the new
	version 3 psydat file format

	* mpsy_adapt_1up_2down.m: and all other files
	* mpsy_adapt_*.m:
	Fix of a typo in the comment when a LOWER reversal is detected

	* mpsy_check.m: updated check for psydat file version from
	version 2 to new version 3.

2014-04-02  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_maxreversal_check.m (mpsy_maxreversal_check):  new file
	to check for value of M.MAXREVERSAL:  In older version up to
	2.5 only the upper reversals were counted.  From version 2.6
	and higher, both upper and lower reversals are counted, and
	M.MAXREVERSAL should therefore be twice as large as earlier.
	This file serves as an automatic check for smaller value of
	M.MAXREVERSAL (<=4) in older psylab experiments.  

2014-03-18  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_det_reversal_adj_step.m: File is made obsolete.  Its
	functionality is now placed into each separate file with an
	adaptive method, like mpsy_adapt_1up_1down.m,
	mpsy_adapt_1up_2down.m, mpys_adapt_uwud.m, etc.  The reasons are:
	1) The code looks much easier in the non-general form.
	2) In case of the uwud-method, the "unsure" answer is treated
	as a "wrong" answer (go up) but is not marked by 0 or 1 in
	array M.ANSWERS
	3) From version 2.6, the counting of reversals has changed to
	count both uppper AND lower reversals.  This is now also done in
	each file for adaptive procedures.

2014-02-20  Martin Hansen  <psylab AT jade-hs.de>

	* display_psydat.m (display_psydat): fixed error for the case
	npar==1 :
npar     = size(y,2)-4;
if npar == 1,
  % hm... only 1 parameter.  Add one fake-column for par2, being
  % only zero.  this will then be the one-and-only value of par2,
  % thus yielding just 1 "unique" value, i.e. uni_par2==1, see below
  y = [y 0*y(:,end) ];
  % despite only 1 parameter, 'npar' must reflect size of matrix y
  npar = npar + 1;
  param_perm = (1:npar);   % THIS line is new
end

	* display_psydat_raw.m (display_psydat_raw):
	fix of same error as in display_psydat.m


2013-09-09  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_plot_feedback.m: changed look of plot:
	familiarization	phase is now plotted in gray and measurement
	phase in black.

2013-04-10  Martin Hansen  <psylab AT jade-hs.de>

	changed affiliation/email of all files into jade-hs.de

2012-11-23  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_intrlv_match_main.m: new file for interleaved matching
	experiments
	* mpsy_match_main.m: the earlier version used a hard-coded
	1up_1down method as the only method of matching experiment.
	This has been changed, such that the adaptive method has to be
	specified now.  This is needed in cases where, e.g., the 29.3%
	point or/and the 70.7% point for a percept like "A is louder than
	B" is desired.  Omitted specification of the method will now lead
	to an error.

2011-11-01  Martin Hansen  <psylab AT jade-hs.de>

	* mpsy_msound_present.m:
	new functionality to use msound for playback (at the
	users option).  Originally this was to overcome errors of
	"sound" in Matlab-2011 Versions.  Also, visual interval
	indication is now always possible in combination with msound,
	which is useful on non-win PCs.

	* mpsy_afc_present.m:
	added new msound functionality

	* mpsy_visual_interval_check.m:
	new file to perform checks for correct (equal) lengths of
	m_test and m_ref which are required for visual interval
	indication.  This script is now called from
	mspsy_visual_interval_indicator and from mpsy_msound_present

	* mpsy_init_run.m:
	reflected new feature of msound-playback

	* mpsy_2afc_gui.m, mpsy_2auc_gui.m, ..., mpsy_4auc_gui.m:
	all GUI answer buttons handles have changed from individual
	variable names (afc_b1, afc_b2, etc.) to an array afc_but(1),
	afc_but(2), etc.

	* mpsy_visual_interval_indicator.m:
	reflecting change of answer button handle names made this
	script simpler.

	* mpsy_check.m:
	reflected the change of answer button handle names


2009-11-10  Martin Hansen  <psylab AT fh-oldenburg.de>

	* mpsy_intrlv_afc_main.m: same change regarding reversed order
	as in mpsy_afc_main.m and mpsy_match_main.m

	Also, applied same change in quit-mechanism as for
	mpsy_match_main.m and mpsy_afc_main.m, see 2009-11-05.

	* mpsy_match_main.m: same change as in mpsy_afc_main.m

	* mpsy_afc_main.m: new reversed order of


mpsy_check.m
% Usage: y = mpsy_check()
%
% performs a check on certain PSYLAB variables in order to ensure that
% vital variables are properly set by the user and their values are of
% acceptable type/content prior to starting an experimental run
% 
%   input:   (none), acts on a set of global variables
%  output:   (none)
%
% Copyright (C) 2005 by Martin Hansen, Fachhochschule OOW
% Author :  Martin Hansen <psylab AT jade-hs.de>
% Date   :  08 Nov 2005
% Updated:  <15 Apr 2016 16:55, martin>
% Updated:  <23 Okt 2006 00:05, hansen>
% Updated:  < 8 Nov 2005 22:10, mh>

%% This file is part of PSYLAB, a collection of scripts for
%% designing and controlling interactive psychoacoustical listening
%% experiments.  
%% This file is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published
%% by the Free Software Foundation; either version 2 of the License, 
%% or (at your option) any later version.  See the GNU General
%% Public License for more details:  http://www.gnu.org/licenses/gpl
%% This file is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


% check for experiment using old psylab version via existence of
% old-style variable M_PARAM 
if exist('M_PARAM'),
  fprintf('*** You seem to try to run an experiment made for psylab version 1,\n');
  fprintf('*** while your psylab appears to be version %s. \n', mpsy_version);
  fprintf('*** You will probably need to change one or the other to a matching version.\n');
  warning('version mismatch of psylab version and your experiment files');
end

% check for experiment using old style definition of adaptive procedure
if isfield(M, 'ADAPT_N_UP') & ~isfield(M, 'ADAPT_METHOD'),
  fprintf('*** You seem to try to run an experiment made for psylab version prior to 2.3,\n');
  fprintf('*** while your psylab appears to be version %s. \n', mpsy_version);
  fprintf('*** You will probably need to change one or the other to a matching version.\n');a
  warning('version mismatch of psylab version and your experiment files');
end

% field variables that must contain a string
mpsy_field_must_be_string = {'SNAME', 'VARNAME', 'VARUNIT'};

% field variables that must contain a numeric scalar value
mpsy_field_must_be_scalar = {'VAR', 'STEP', 'MINSTEP', 'NUM_PARAMS'};
% field variables that must contain a numeric value (scalar or vector)
mpsy_field_must_be_numeric = [mpsy_field_must_be_scalar, 'PARAM'];

% these field variables need to be cells (of stings), EVEN if only 1 element exist
mpsy_field_must_be_cell = {'PARAMNAME', 'PARAMUNIT'};


mpsy_field_must_not_be_empty = [mpsy_field_must_be_string mpsy_field_must_be_numeric];

% check for empty variables
for tmp = mpsy_field_must_not_be_empty,
  tmp_varname = char(tmp);
 
  % check for existence first:
  if ~isfield(M, tmp_varname), 
    error('psylab-variable "M.%s" must exist and must not be empty!  \n',  tmp_varname);
  end
  % then check for emptiness:
  if isempty(getfield(M, tmp_varname)),
    error('psylab-variable "M.%s" must not be left empty! \n',  tmp_varname);
  end
end

% check for numeric type variables
for tmp = mpsy_field_must_be_numeric,
  tmp_varname = char(tmp);
  if ~isnumeric(getfield(M, tmp_varname)),
    error('content of psylab-variable "M.%s" must be a numeric value!  \n',  tmp_varname);
  end
end

% check for string type variables
for tmp = mpsy_field_must_be_string,
  tmp_varname = char(tmp);
  if ~ischar(getfield(M, tmp_varname)),
    error('psylab-variable "M.%s" must be a string!  \n', tmp_varname);
  end
  if strfind(getfield(M, tmp_varname), ' '),
    error('content of psylab-variable "M.%s" must not contain blank spaces!  ', tmp_varname);
  end
end

% check for being scalar variables
for tmp = mpsy_field_must_be_scalar,
  tmp_varname = char(tmp);
  if ~isscalar(getfield(M, tmp_varname)),
    error('psylab-variable "M.%s" must be a scalar!  \n', tmp_varname);
  end
end


% check for correct lengths of M.PARAM* variables
if length(M.PARAM) ~= M.NUM_PARAMS,
  error('length of M.PARAM must match value of M.NUM_PARAMS ');
end
if length(M.PARAMNAME) ~= M.NUM_PARAMS,
  error('length of M.PARAMNAME must match value of M.NUM_PARAMS ');
end
if length(M.PARAMUNIT) ~= M.NUM_PARAMS,
  error('length of M.PARAMUNIT must match value of M.NUM_PARAMS ');
end

% check for cellstring type variables
for tmp = mpsy_field_must_be_cell,
  tmp_varname = char(tmp);
  tmp_val = getfield(M,tmp_varname);
  if ~iscellstr(tmp_val),
    error('psylab-variable "%s" must be a cell(string)!  \n', tmp_varname);
  end
  for k=1:M.NUM_PARAMS
    if strfind(char(tmp_val(k)), ' '),
      error('content of psylab-variable "%s(%d)" must not contain blank spaces!  \n', tmp_varname, k);
    end
    if length(char(tmp_val(k))) == 0,
      error('psylab-variable "%s(%d)" must not be empty!  \n', tmp_varname, k);
    end
  end
end


%% --------------------------------------------------
%% check type of experiment (e.g. n-AFC or matching) 
clear tmp_type; 
tmp_type(1) = isfield(M, 'NAFC');
tmp_type(2) = isfield(M, 'MATCH_ORDER');
tmp_type(3) = isfield(M, 'N_RECOGNITION');
if sum(tmp_type) ~= 1,
  error('exactly one of either M.NAFC or M.MATCH_ORDER or M.N_RECOGNITION must be set correctly');
end

%% --------------------------------------------------
%% check correct sign of M.STEP and M.MINSTEP
if sign(M.STEP) ~= sign(M.MINSTEP),
  error('M.STEP and M.MINSTEP must have the same sign:  both positive for regular experiments, both negative for reversed-up-and-down experiment');
end



%% --------------------------------------------------
% check proper value of M.VISUAL_INDICATOR
if M.VISUAL_INDICATOR & ~M.USE_GUI,
  fprintf('\n*** INFO: the VISUAL_INDICATOR feature can only be used when using a GUI\n');
  warning('M.VISUAL_INDICATOR has been set to 0.');
  pause(2);
  M.VISUAL_INDICATOR = 0;
end



%% --------------------------------------------------
% check version of psydat file, create correct header line 
% if file doesn't exist yet
M.PSYDAT_VERSION = mpsy_check_psydat_file( ['psydat_' M.SNAME] );
if M.PSYDAT_VERSION < 3,
  fprintf('\n\n*** You are using psylab with psydat file format version 3\n');
  fprintf('*** but you seem to have an existing psydat file in the current \n');
  fprintf('*** directory which is in version %d format.  \n', M.PSYDAT_VERSION); 
  fprintf('*** While your data would be saved without any loss, a mix of different\n')
  fprintf('*** data formats in the same psydat file would destroy its automatic\n')
  fprintf('*** readibility.\n');  
  fprintf('*** Possible solution: Store your old psydat file in a different directory \n');  
  fprintf('*** or move your experiment to another directory now and restart it there.\n')
  fprintf('*** Your psydat file in version 2 can be read and processed by the function \n')
  fprintf('*** "read_psydat_v2", while "read_psydat" will use psydat version 3\n')
  fprintf('*** \n')
  error('file "%s" in current directory seems to be in psydat format version %d', ...
        ['psydat_' M.SNAME], M.PSYDAT_VERSION );
end



%% --------------------------------------------------
%% ensure existence of pre-signal, quiet-signal, and post-signal
%% - if not yet existent, thencreate empty signals
%% - if existent but a scalar value, then generate silence signals
%
% For the generation of silence signals, we need to know the number
% of channels (mono, stereo, ...).  In order to determine that
% number of channels, run the user-script once here and check the
% size of the signal m_test which will then have been generated:
eval([M.EXPNAME 'user']); 
numchannels = size(m_test, 2);

if exist('m_quiet')~=1,   
  m_quiet   = [];  
else
  if isscalar(m_quiet),
    m_quiet = zeros(round(m_quiet*M.FS), numchannels);
  end
end
if exist('m_presig')~=1,  
  m_presig  = [];  
else
  if isscalar(m_presig),
    m_presig = zeros(round(m_presig*M.FS), numchannels);
  end
end
if exist('m_postsig')~=1, 
  m_postsig = [];  
else
  if isscalar(m_postsig),
    m_postsig = zeros(round(m_postsig*M.FS), numchannels);
  end
end


% $$$  this was a test - not working properly.  
% $$$ %% --------------------------------------------------
% $$$ %% (re)create gui 
% $$$ if M.USE_GUI,
% $$$   htmp = findobj('Tag', 'psylab_answer_gui');
% $$$   if ~isempty(htmp),
% $$$     % Delete the already existing GUI.  This is necessary
% $$$     % especially in (some rare) cases where a matching experiment
% $$$     % follows an AFC experiment, or vice versa. 
% $$$     delete(htmp)
% $$$   end
% $$$   
% $$$   % create a fresh GUI.  This avoids possible problems with cleared
% $$$   % (handle) variables for the GUI objects
% $$$   if isfield(M, 'NAFC') & M.NAFC >= 2,
% $$$     % hm, seems like we have an n-AFC experiment
% $$$     eval( ['mpsy_' num2str(M.NAFC) 'afc_gui;'] );
% $$$   elseif isfield(M, 'MATCH_ORDER') & M.MATCH_ORDER >= 0,
% $$$     % seems like we have a matching experiment
% $$$     mpsy_up_down_gui;
% $$$   end
% $$$ end


%% --------------------------------------------------
%% create gui or get the handle for the afc gui 
if M.USE_GUI,
  htmp = findobj('Tag', 'psylab_answer_gui');
  if isempty(htmp),
    % in case of first run or in case the answer-gui got cleared/closed 
    % unintentionally by the test subject:  create a fresh one.  
    if isfield(M, 'NAFC') & M.NAFC >= 2,
      if strcmp(M.ADAPT_METHOD, 'uwud')
        % seems like we have an unforced choice n-AUC experiment
        eval( ['mpsy_' num2str(M.NAFC) 'auc_gui;'] );
      else
        % seems like we have an n-AFC experiment
        eval( ['mpsy_' num2str(M.NAFC) 'afc_gui;'] );
      end
      
    elseif isfield(M, 'MATCH_ORDER') & M.MATCH_ORDER >= 0,
      % seems like we have a matching experiment
      mpsy_up_down_gui;
    end
    
  else 
    % the answer-gui is already/still there, 
    % so just place it visibly in the foreground 
    figure(htmp);
    % get the handles for the feedback/info text in case they got cleared
    afc_fb = findobj('Tag','psylab_feedback'); 
    afc_info = findobj('Tag','psylab_info'); 
    
    if isfield(M, 'NAFC')
      % in case of ACF experiment, also get the handles for the interval 
      % buttons, as they are needed by mpsy_visual_interval_indicator.m   
      for kk=1:M.NAFC
        afc_but(kk) = findobj('Tag',['afc_answer' num2str(kk)] );
    
        % $$$       afc_but(1) = findobj('Tag','afc_answer1');
        % $$$       afc_but(2) = findobj('Tag','afc_answer2');
        % $$$       afc_but(3) = findobj('Tag','afc_answer3');
        % $$$       afc_but(4) = findobj('Tag','afc_answer4');
      end
    else
      % in case of a matching experiment, get the two button handles
      afc_but(1) = findobj('Tag','ButDOWN');
      afc_but(2) = findobj('Tag','ButUP');
    end
    
  end
end




% End of file:  mpsy_check.m

% Local Variables:
% time-stamp-pattern: "40/Updated:  <%2d %3b %:y %02H:%02M, %u>"
% End:
